base/meta.jl

```diff
@@ -7,17 +7,66 @@ module Meta
 
 export quot,
        isexpr,
+       propagate_sourceloc!,
        show_sexpr
 
+"""
+    quot(ex)
+
+Return `Expr(:quote, ex)`.  This is useful for interpolating symbols into an
+AST as symbol literals rather than variable names.
+"""
 quot(ex) = Expr(:quote, ex)
 
+"""
+    isexpr(ex, head [, n])
+
+Return true if `ex` is an `Expr` with the given type `head` and optionally that
+the argument list is of length `n`.  `head` may be a `Symbol` or collection of
+`Symbol`s.  For example, to check that a macro was passed a function call
+expression, you might use `isexpr(ex, :call)`.
+"""
 isexpr(ex::Expr, head)          = ex.head === head
 isexpr(ex::Expr, heads::Union{Set,Vector,Tuple}) = in(ex.head, heads)
 isexpr(ex,       head)          = false
 
 isexpr(ex,       head, n::Int)  = isexpr(ex, head) && length(ex.args) == n
 
 
+"""
+    propagate_sourceloc!(location, expr)
+
+Overwrite the caller source location embedded in the macro call AST for all
+macro calls inside `expr`, and return the resulting AST.  This is most useful
+to make the `@__LINE__` and `@__FILE__` macros return the location where a
+parent macro was invoked.  For example:
+
+```
+macro example()
+    propagate_sourceloc!(__source__,
+        quote
+            @__LINE__
```

Comment:
That is, if package `A` provides the macro

```julia
macro a()
    quote
        some_complicated_expression($(__source__.line))
    end
end
```

And package `B` wishes to wrap it - 

```julia
macro b()
    quote
        # ...
        @a
    end
end
```

while preserving reporting of the source location, you either need a tool a bit like this, or to manually dig into the AST generated in `@b` and rewrite the source location of the macro invocations contained within.